src/rgw/driver/d4n/rgw_sal_d4n.cc

```diff
@@ -306,104 +506,541 @@ std::unique_ptr<Object::DeleteOp> D4NFilterObject::get_delete_op()
+
+  if (!last_part && bl.length() <= rgw_get_obj_max_req_size) {
+    auto r = client_cb->handle_data(bl, bl_ofs, bl_len);
+
+    if (r < 0) {
+      return r;
+    }
   }
 
-  int delObjReturn = source->filter->get_d4n_cache()->delObject(source->get_key().get_oid());
+  //Accumulating data from backend store into rgw_get_obj_max_req_size sized chunks and then writing to cache
+  if (write_to_cache) {
+    rgw::d4n::CacheBlock block, existing_block;
+    rgw::d4n::BlockDirectory* blockDir = source->driver->get_block_dir();
+    block.hostsList.push_back(blockDir->cct->_conf->rgw_local_cache_address); 
+    block.cacheObj.objName = source->get_key().get_oid();
+    block.cacheObj.bucketName = source->get_bucket()->get_name();
+    std::stringstream s;
+    utime_t ut(source->get_mtime());
+    ut.gmtime(s);
+    block.cacheObj.creationTime = s.str(); 
+    block.cacheObj.dirty = false;
+
+    //populating fields needed for building directory index
+    existing_block.cacheObj.objName = block.cacheObj.objName;
+    existing_block.cacheObj.bucketName = block.cacheObj.bucketName;
+    Attrs attrs; // empty attrs for cache sets
+    std::string version = source->get_object_version();
+    std::string prefix = source->get_prefix();
+
+    ldpp_dout(dpp, 20) << __func__ << ": version stored in update method is: " << version << dendl;
+
+    if (bl.length() > 0 && last_part) { // if bl = bl_rem has data and this is the last part, write it to cache
+      std::string oid = prefix + "_" + std::to_string(ofs) + "_" + std::to_string(bl_len);
+      if (!filter->get_policy_driver()->get_cache_policy()->exist_key(oid)) {
+        block.blockID = ofs;
+        block.size = bl.length();
+        block.version = version;
+        auto ret = filter->get_policy_driver()->get_cache_policy()->eviction(dpp, block.size, *y);
+        if (ret == 0) {
+          ret = filter->get_cache_driver()->put(dpp, oid, bl, bl.length(), attrs, *y);
+          if (ret == 0) {
+            filter->get_policy_driver()->get_cache_policy()->update(dpp, oid, ofs, bl.length(), version, *y);
+
+	    /* Store block in directory */
+            if (!blockDir->exist_key(&block, *y)) {
+              if (blockDir->set(&block, *y) < 0) //should we revert previous steps if this step fails?
+		ldpp_dout(dpp, 10) << "D4NFilterObject::D4NFilterReadOp::D4NFilterGetCB::" << __func__ << "(): BlockDirectory set method failed." << dendl;
+            } else {
+              existing_block.blockID = block.blockID;
+              existing_block.size = block.size;
+              if (blockDir->get(&existing_block, *y) < 0) {
+                ldpp_dout(dpp, 10) << "Failed to fetch existing block for: " << existing_block.cacheObj.objName << " blockID: " << existing_block.blockID << " block size: " << existing_block.size << dendl;
+              } else {
+                if (existing_block.version != block.version) {
+                  if (blockDir->del(&existing_block, *y) < 0) //delete existing block
```

Comment:
Do we have to delete the block here or can we simply set it normally? Setting it will overwrite all the values, which is what we want.

Reply:
We don't need to delete it, which is what I have described in the other PR of yours, related to reducing the number of redis calls.

Reply:
I must have missed where you said that, but that makes sense. I'll remove these `delete` calls in my PR